src/WebJobs.Script.WebHost/WebScriptHostExceptionHandler.cs

```diff
@@ -26,21 +27,34 @@ public void Initialize(JobHost host)
         {
         }
 
-        public Task OnTimeoutExceptionAsync(ExceptionDispatchInfo exceptionInfo, TimeSpan timeoutGracePeriod)
+        public async Task OnTimeoutExceptionAsync(ExceptionDispatchInfo exceptionInfo, TimeSpan timeoutGracePeriod)
         {
+            FunctionTimeoutException timeoutException = exceptionInfo.SourceException as FunctionTimeoutException;
+
+            if (timeoutException?.FunctionTask != null)
+            {
+                // We may double the timeoutGracePeriod here by first waiting to see if the iniital
+                // function task that started the exception has completed.
+                Task completedTask = await Task.WhenAny(timeoutException.FunctionTask, Task.Delay(timeoutGracePeriod));
+
+                // If the function task has completed, simply return. We've already logged the timeout.
+                if (completedTask == timeoutException.FunctionTask)
+                {
+                    return;
+                }
+            }
+
             LogErrorAndFlush("A function timeout has occurred. Host is shutting down.", exceptionInfo.SourceException);
 
             // We can't wait on this as it may cause a deadlock if the timeout was fired
             // by a Listener that cannot stop until it has completed.
-            Task taskIgnore = _manager.StopAsync();
+            Task taskIgnoreStop = _manager.StopAsync();
```

Comment:
FxCop doesn't like it if you don't await `async` methods. This was the only way I could get it to compile. I noticed that we do it elsewhere as well, which is how I figured this out: https://github.com/Azure/azure-webjobs-sdk-script/blob/dev/src/WebJobs.Script/Host/ScriptHostManager.cs#L126.

I previously had a global suppression for this (as we do for the link above) -- but I removed it and it still compiles. It looks like if you have an `async` method you're allowed to have unused local `Task` variables like this (I just tested it by removing `async`). This method used to not be `async` so I previously needed the suppression.